druid-shell/src/mac/menu.rs

```diff
@@ -13,17 +13,106 @@
 // limitations under the License.
 
 //! macOS implementation of menus.
+use lazy_static::lazy_static;
+use objc::runtime::{Class, Object, Sel};
+use objc::declare::ClassDecl;
+use cocoa::base::{id, nil, BOOL, NO, YES};
+use cocoa::foundation::{NSAutoreleasePool, NSPoint, NSRect, NSSize, NSString};
+use cocoa::appkit::{NSMenu,NSMenuItem, NSApplication};
 
-pub struct Menu;
+struct MenuItemProxyClass(*const Class);
+unsafe impl Sync for MenuItemProxyClass {}
 
+lazy_static! {
+    static ref MENU_ITEM_PROXY_CLASS: MenuItemProxyClass = {
+        unsafe {
+            let mut decl = 
+                ClassDecl::new("DruidMenuItemProxy", class!(NSObject))
+                .expect("Menu class defined");
+            decl.add_ivar::<u32>("menu_id");
+            decl.add_method(
+                sel!(trigger),
+                trigger as extern "C" fn (&Object, Sel)
+            );
+            extern "C" fn trigger(this: &Object, _: Sel) {
+                unsafe {
+                    let menu_id: u32 = *this.get_ivar("menu_id");
+                    println!("triggered menu item with id {}", menu_id);
+                }
+            }
+            MenuItemProxyClass(decl.register())
+        }
+    };
+}
+pub struct Menu {
+    pub menu: id,
+}
+fn make_basic_menu_item(id: u32, text: &str, key: &str) -> id {
+    unsafe {
+        NSMenuItem
+            ::alloc(nil)
+            .initWithTitle_action_keyEquivalent_(
+                NSString::alloc(nil).init_str(text),
+                sel!(trigger),
+                NSString::alloc(nil).init_str(key)
+            ).autorelease()
+    }
+}
+
+fn make_proxy(id: u32) -> id {
+    unsafe {
+        let mut target: id = msg_send![MENU_ITEM_PROXY_CLASS.0, alloc];
+        (*target).set_ivar("menu_id", id);
+        target.autorelease()
+    }
+}
+fn make_menu_item(id: u32, text: &str, key: &str) -> id {
+    unsafe {
```

Comment:
I've talked with @raphlinus before about this and I think we have different opinions. I prefer to wrap the absolute minimum in unsafe so that debugging is easier. With the code as it is you'd have to inspect `make_proxy` only to realize it's not actually unsafe. In this case that's not a huge deal since it's right above, but if this function were in another crate, you'd have to do some digging that could be avoided if `unsafe` were used a bit less liberally.

Reply:
I'm not sure I fully understand. The reason I've done it this way is because basically if you're going to touch cocoa integration, you're fundamentally in unsafe territory because of the need to interop with the objectionable c runtime. I don't see a good way of avoiding unsafe in the general case with what we're doing, so i haven't worried too hard about it.

Reply:
I'm refering to the scope of the `unsafe` block. You could move the use of the `unsafe` keyword down one line and the code would still compile. I'm arguing that this is the right thing to do because it properly highlights which code is actually unsafe in that function.